### PR TITLE
Piano Roll and Automation editor grid redesign (w/ @BaraMGB)

### LINKED ISSUE #3019: Piano Roll Grid Contrast
I'm trying to adjust to the new theme and find it particularly hard to locate quarter notes.  This may be a matter of an old dog trying to learn a new trick, but I continuously find myself misjudging where to place notes.

At a glance, the major difference seems to be the grid contrast on the background.  Is anyone else experiencing this same problem?  If this is an overwhelming "no", we can close this out as invalid.

| Old | New |
| --- | --- |
| screen shot 2016-09-04 at 4 34 07 pm | screen shot 2016-09-04 at 4 36 10 pm |

### PR DESCRIPTION
**Pull request by Umcaruje**
Fixes #3019 and #2308

Thank you very much @BaraMGB for collaborating on this Pull Request with me and helping me out with my problems.

screenshot from 2016-10-03 20 34 37
screenshot from 2016-10-03 20 34 18

The grid is ligher, the background alternates shades. I also did a general CSS cleanup, fixed the indentation and stuff.

Feedback and testing welcome.

### CODE REVIEW

**Comment by BaraMGB** [comment]
That looks really great, now.

**Comment by simonvanderveldt** [comment]
Could it also solve this one (or include a specific fix for it)? #2832

**Comment by Umcaruje** [comment]
>Could it also solve this one (or include a specific fix for it)? #2832

To allow seperate colors for different lines? seems like a overkill to me.

**Comment by Umcaruje** [comment]
Conflicts resolved, rebased.

**Comment by Umcaruje** [comment]
@musikBear your feedback would be valuable on this.

**Comment by BaraMGB** [comment]
Could it be worth to shade horizontal lines which represent a black key? I noticed that in other programs.

**Comment by musikBear** [comment]
@Umcaruje Very good! Also in respect to the vertical grid-lines. I think this will be pleasant to work with
:champagne: :+1:

**Comment by BaraMGB** [comment]
2016-10-06-141647_1280x800_scrot

@Umcaruje I've got a green line on top of the shading rect. :worried:

File: src/gui/editors/PianoRoll.cpp [hunk 1]
```diff
@@ -2863,6 +2870,22 @@ void PianoRoll::paintEvent(QPaintEvent * pe )
...
+	int barCount = m_currentPosition / MidiTime::ticksPerTact();
+	int leftBars = m_currentPosition / m_ppt;
+
+	p.setBrush( backgroundShade() );
+
+	for ( int x = WHITE_KEY_WIDTH; x < width() + m_currentPosition; x += m_ppt, ++barCount )
+	{
+		if ( (barCount + leftBars)  % 2 != 0 )
+		{
+				p.drawRect( x - m_currentPosition, PR_TOP_MARGIN, m_ppt,
```
Comment (BaraMGB) [inline src/gui/editors/PianoRoll.cpp]: Okay, I've got a green line on top of the shading rect:
2016-10-06-143342_1920x1080_scrot

What helps is:
```
p.fillRect( x - m_currentPosition, PR_TOP_MARGIN, m_ppt,
					height() - ( PR_BOTTOM_MARGIN + PR_TOP_MARGIN ), backgroundShade() );
```

by removing the : ```p.setBrush( backgroundShade() );``` of course. 


```

**Comment by Umcaruje** [comment]
Good catch @BaraMGB, will fix asap.

Glad you like it @musikbear. Can we close out the issue where its asked
that the tick lines have a different color? Or do you think this still does
not help in that sense

**Comment by musikBear** [comment]
@Umcaruje afaik, you have implemented css option for grid, and my poor vision can not be a hallmark for the design. I can fiddle with the css, and the current design you have here is clearly much better. The vertical bar-dividers are still a bit difficult to see, but yes. If noone else feel the same about bar-dividers, then this is a fine design :+1: 
[ I am not sure i ever really got to explain the triplet-line idea for bar-dividers, so i will just try with one more drawing, because there was never different colors in that idea, just two additional lines on each side of the current bar-divider-line, so in total 3 lines

I have tried to **explain** the tripple-line bar-divider WITH color, but that just to explain the idea
tripletdivider

**Comment by Umcaruje** [comment]
@BaraMGB fixed it.

**Comment by Umcaruje** [comment]
@musikBear here's how the grid looks with your proposal. I don't like it too much, but not entirely against it either.
screenshot from 2016-10-07 21 43 25

**Comment by simonvanderveldt** [comment]
Hmm, the 3 lines might be a bit excessive.

But on the other hand, especially in the automation editor the distinction between the bar and and tick lines is pretty minimal (like in your screenshot https://cloud.githubusercontent.com/assets/6282045/19049224/d129f1c0-89a9-11e6-82cb-1d2cb7766374.png).
It would be nice of those lines were a bit easier to distinguish from eachother.

**Comment by BaraMGB** [comment]
If the discussion is over, this one can be merged. @musikBear @simonvanderveldt ?

**Comment by simonvanderveldt** [comment]
@BaraMGB I just tried this branch locally and for me the extra shading of every 2nd bar makes LMMS feel a bit "busy" (hope you know what I mean). I definitely prefer it without the shading.

I do think a better visible **grid** is a necessity though. This goes for both the normal horizontal (notes) and vertical (ticks) lines and especially for the "major" ticks.
For me the bar start/end is clearly visible, but the major ticks like 1.2, 1.3, etc aren't.

Best example of what works really well for me/I really like is this:
https://www.image-line.com/support/FLHelp/html/img_shot/pianoroll_draghandle.png

@Umcaruje  Any clue if this will clash with darker horizontal lines for the black keys vs lighter horizontal lines for the white keys?

**Comment by Umcaruje** [comment]
I'm going to merge this in a couple of hours if there are no objections.

**Comment by simonvanderveldt** [comment]
> I'm going to merge this in a couple of hours if there are no objections.

I think we should address the root cause that currently makes it difficult to distinguish bars (and ticks) from each other, which IMHO is that the grid's visibility isn't good enough. This is pretty much as posted in #3019 

I'm not a fan of adding extra stuff (in this case visual overhead) to fix issues whilst the root issue isn't solved. I do see that #2308 specifically requests highlighting for odd bars, but no one asked why that request what made.

**Comment by Umcaruje** [comment]
Ok then, I'll work on this more :+1:

**Comment by musikBear** [comment]
@Umcaruje Sorry i have been 'Off' for a week, so i have not been following your work. I have to say that i kind of like the image with the triplet line, but i also understand that you feel it does not suit your overall design. Design and 'taste' is darn difficult :) Very good work, though :+1:

**Comment by Umcaruje** [comment]
Update: 
image

Figured out the grid finally, now time to do the same in the Automation editor, and this will be ready for review.

**Comment by musikBear** [comment]
> Figured out the grid finally, now time to do the same in the Automation editor, and this will be ready for review.

@Umcaruje -is alternate bars shading no longer considered?

**Comment by Umcaruje** [comment]
>@Umcaruje -is alternate bars shading no longer considered?

If you look at my screenshot, it's still there, although I might add a bit more contrast to it.

**Comment by tresf** [comment]
> If you look at my screenshot, it's still there, although I might add a bit more contrast to it.

No, it's not remotely visible in the screenshot on any of my monitors.

**Comment by Umcaruje** [comment]
Note to self: `rgba(255, 255, 255, 4)` is not an acceptable color for shading

**Comment by BaraMGB** [comment]
I can see it. 😳 😏

**Comment by Umcaruje** [comment]
Update: Automation editor has the same grid now, and shading is more visible. 

Waiting for color guidelines from @RebeccaDeField and this can be tested and merged.
image
image

**Comment by musikBear** [comment]
> Update: Automation editor has the same grid now, and shading is more visible.
@Umcaruje it is visible now, but inside the blue area in automation, i cant see it  ( -It is ofcause also because of the graduated alpha, which i know is easy to disable :)
Is the intensity of the bar shading also modifiable in css?
The me and other bats, can just do that, and i will not comment the faint default grid again :)

**Comment by Umcaruje** [comment]
>Is the intensity of the bar shading also modifiable in css?

Yes, absolutely everything is modifiable in css. You can easily change the opacity of the shading part.

**Comment by tresf** [comment]
@musikBear has a good point.  This may be worthy running through a colorblind test.  Here's some numbers... 

> an all-boys school in the Home Counties of England with 1000 pupils would have approximately 100 colour deficient students. 12-13 would be deuteranopes, 12-13 would be protanopes, 12-13 would have a form of protanomaly and 62 would have a form of deuteranomaly. About half of those with an anomalous condition would have a moderate to severe form of deficiency.

**Comment by tresf** [comment]
That said, I love it. :)

**Comment by Umcaruje** [comment]
After some help from @RebeccaDeField this is the current look: 
screenshot from 2017-02-13 00 37 53
screenshot from 2017-02-13 00 37 39

This PR is finally ready for merging and review.

**Comment by Umcaruje** [comment]
@tresf @musikBear @vlad1777d If this is looking good, I'll merge.

**Comment by musikBear** [comment]
@Umcaruje When you told me that i can work with it in CSS, im absolutely satisfied! Great job! 🎆

**Comment by tresf** [comment]
@Umcaruje I have not reviewed the code, but the UI without a doubt fixes #3019. 👍

**Comment by vlad1777d** [comment]
@Umcaruje , It looks great, especially violent automation line =)

**Comment by vlad1777d** [comment]
@Umcaruje , sorry that I write not in appropriate place, but I didn't found any IRC channels or something like this to ask question,
can I change colors of automation editor's grid without rebuilding LMMS ?

Seems, it's not so clear as Piano Roll's grid.

**Comment by zonkmachine** [comment]
> I didn't found any IRC channels or something like this to ask question,

Here: https://discord.gg/3sc5su7
Especially the channel named `#support`

**Comment by musikBear** [comment]
> Here: <discord>

Well yes, but any explanation done in discord, is lost after the discord session. So the time used in discord to explain a feature, benefits exactly *one* person.  
Written down, it could benefit n persons. 🙊

**Comment by Spekular** [comment]
@musikBear nope, Discord has persistent chat history. Besides, mentioning "consensus on Discord is ___" *is* writing it down for the benefit of those on github.

**Comment by Umcaruje** [comment]
>@Umcaruje , sorry that I write not in appropriate place, but I didn't found any IRC channels or something like this to ask question,
can I change colors of automation editor's grid without rebuilding LMMS ?

@vlad1777d yeah, you can. Just edit the `qproperty-lineColor, qproperty-beatLineColor, qproperty-barLineColor` css rules under `AutomationEditor`. 

If you come up with a better colored grid, don't be hesitant to share, I'm always open for improvement on the gui, and your mockups made this PR happen :)

**Comment by musikBear** [comment]
> Discord has persistent chat history

Searchable? That would render my thought to void, but is it? ..anywitch `oftopic` so ..

**Comment by Spekular** [comment]
@musikBear searchable, and with great filtering:
https://support.discordapp.com/hc/en-us/articles/115000468588-Using-Search

**Pull request merged**